### PR TITLE
Fix(metrics): decorated class methods cannot access `this`

### LINKED ISSUE #1057: Bug: decorated class methods cannot access `this` in Metrics
## Bug description

A long form description of this issue is available in #1056. This is an issue that affects all three Powertools utilities because they all share similar implementation for the decorators.

---

When using Powertools in with a Class and when decorating a method, the decorated method does not have access to the correct value of `this` and so it's not able to call other methods nor read attributes from its same class.

This is due to the fact that the decorator implementation uses an `() => ()` arrow function instead of a regular `function () {}` one. This mixes the value of `this` which then causes the wrong value to be passed.

Additionally there's a missing detail in the documentation of the Class instrumentation that prevents compounds the issue described above.

### Expected Behavior

Decorated methods have access to the correct value of `this` and is able to call methods or access attributes/properties.

### Current Behavior

See "Steps to Reproduce" section.

### Possible Solution

Apply the same changes done to Tracer in #1055 

### Steps to Reproduce

```ts
import {
  LambdaInterface,
  SyncHandler,
  AsyncHandler,
} from "@aws-lambda-powertools/commons";
import { Metrics } from "@aws-lambda-powertools/metrics";

const metrics = new Metrics({ serviceName: "serverlessAirline" });

class Lambda implements LambdaInterface {
  greeting: string;

  constructor(message: string) {
    this.greeting = message;
  }

  public getGreeting(): string {
    /* ... */
    return `Hello ${this.greeting}`;
  }

  @metrics.logMetrics({ captureColdStartMetric: true })
  public handler(_event: any, _context: any, _callback: any): void {
    console.log(this.getGreeting());
  }
}

export const myFunction = new Lambda("World");
export const handler = myFunction.handler.bind(myFunction);
```

The value logged from `console.log(this.getGreeting());` should be `Hello World` but instead is `Hello undefined`:
Screenshot 2022-08-13 at 22 27 25

### Environment

* **Powertools version used**: 1.1.0
* **Packaging format (Layers, npm)**: all
* **AWS Lambda function runtime:** all
* **Debugging logs:** N/A

### Related issues, RFCs

#1056

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

As described in #1057 the decorator implementation was affected by an issue that prevented the decorated method/class to access other class members as the value of `this` was passed incorrectly.

This PR aims at fixing the issue, creating an unit test case, making minimal changes to the function in the e2e test to verify the change, and update docstrings/docs.

A long form explanation of the issue can be found at #1055, when merged this PR will close #1057. 

### How to verify this change

See checks below the PR, also see the results of the [e2e test run](https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/2855116627).

### Related issues, RFCs

**Issue number:** #1057 

### PR status

***Is this ready for review?:*** YES  
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published* in downstream module
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

N/A

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by dreamorosi** [comment]
Note to reviewers: don't merge this until #1054 is merged.

File: packages/metrics/tests/e2e/basicFeatures.decorator.test.functionCode.ts
Comment (ijemmy) [inline packages/metrics/tests/e2e/basicFeatures.decorator.test.functionCode.ts:3]: Won't this use the published version instead of local version?

**Review by ijemmy — CHANGES_REQUESTED** [review]
Please check the comment.  Want to double check if that is intentional.

Also, similar to #1055, we should also update the `examples/cdk/src/*.Decorator.ts` files. They still don't use .bind() method. This could be done in another PR. But I prefer to have it now or we will forget about this.

File: packages/metrics/tests/e2e/basicFeatures.decorator.test.functionCode.ts
Comment (dreamorosi) [reply packages/metrics/tests/e2e/basicFeatures.decorator.test.functionCode.ts:3]: As far as I understand, given that we are inside the NPM workspace & it uses the only lock file that is present at the root, then no, it should get the local one (see screenshot below):

Comment (dreamorosi) [reply packages/metrics/tests/e2e/basicFeatures.decorator.test.functionCode.ts:3]: image

**Review by ijemmy — APPROVED** [review]

File: docs/core/metrics.md
Comment (saragerion) [inline docs/core/metrics.md:378]: Same as above

Comment (saragerion) [reply docs/core/metrics.md:378]: https://github.com/awslabs/aws-lambda-powertools-typescript/pull/1059/files#r947884023

File: docs/core/metrics.md
Comment (saragerion) [inline docs/core/metrics.md:642]: Same as above: https://github.com/awslabs/aws-lambda-powertools-typescript/pull/1059/files#r947884023

File: packages/metrics/tests/unit/Metrics.test.ts
Comment (saragerion) [inline packages/metrics/tests/unit/Metrics.test.ts:647]: Can you check that it's the metric output?

**Review by saragerion — COMMENTED** [review]
Thanks a lot Andrea!

File: packages/metrics/tests/unit/Metrics.test.ts
Comment (dreamorosi) [reply packages/metrics/tests/unit/Metrics.test.ts:648]: Done! See 3 lines below

**Review by saragerion — APPROVED** [review]

**Review by ijemmy — APPROVED** [review]

**Pull request merged**
